### ISSUE TITLE
Contentful responsive image dimensions can be incorrect

### ISSUE THREAD

**Issue by tremby** [gatsbyjs/gatsby#4539]
## Description

When querying with `resolutions` or `resize` (and probably `sizes`; didn't try), if the requested dimensions are larger than the source image's natural dimensions and upscaling is disabled (as is default), the URL returned for the image points to an image with dimensions smaller than the requested dimensions. However, the `width` and `height` fields in the response say what the dimensions *would* be, had the image been upscaled.

This means the details given in the srcset are incorrect, and also that the `img` tag has the larger dimensions, meaning the image will be upscaled in the browser.

### Steps to reproduce

Query for a Contentful image, requesting dimensions larger than the image. Look at the response URL, width, and height fields. Compare actual image dimensions with those reported.

### Expected result

The dimensions match.

### Actual result

The dimensions do not match.

### Environment

* Gatsby version (`npm list gatsby`): gatsby@1.9.232
* gatsby-cli version (`gatsby --version`): 1.1.41
* Node.js version: v8.9.4
* Operating System: Ubuntu 16.04 x64

### File contents (if changed):

`gatsby-config.js`: nothing special other than contentful source active
`package.json`: nothing special
`gatsby-node.js`: nothing special
`gatsby-browser.js`: not changed
`gatsby-ssr.js`: not changed

**Comment by tremby** [comment]
My workaround for now is this.

Query:

```graphql
...
image {
    title
    resolutions(width: 360) {
        ...GatsbyContentfulResolutions
        aspectRatio
    }
    file {
        details {
            image {
                width
            }
        }
    }
}
...
```

Markup:

```jsx
const width = Math.min(image.resolutions.width, image.file.details.image.width);
const height = width / image.resolutions.aspectRatio;
return <Img resolutions={feature.image.resolutions} style={{width, height}} />;
```

**Comment by KyleAMathews** [comment]
You want to add a test and a fix to this here? https://github.com/gatsbyjs/gatsby/blob/88c0493aef154a1eef364b1cdf94d78c223cb37e/packages/gatsby-source-contentful/src/extend-node-type.js#L50

**Comment by tremby** [comment]
I'm on a really tight deadline right now. I can try when things ease up.

Is the natural size available in that context somewhere?

**Comment by KyleAMathews** [comment]
Yeah, the line I linked to.

**Comment by tremby** [comment]
Oh. I took your message as if that line needed to be fixed, but I guess you're saying this is where to get the dimensions, which need to be used for the fix elsewhere.

**Comment by KyleAMathews** [comment]
Yeah you'd check I'd assume if the args are bigger than than the image size and restrict that

**Comment by tremby** [comment]
@KyleAMathews, I still fully intend to fix this. Can you reopen it, please?

It occurs to me that I'm not sure what the best behaviour would be for the `width` and `height` attributes of the `img` tag. Should they be the size of the actual image coming back (meaning the browser upscales it to what you asked for), or should they be the size that was originally requested?

(The data in `srcset` should obviously be the real dimensions.)

**Comment by KyleAMathews** [comment]
Sure thing!

**Comment by tremby** [comment]
Any thoughts on my question above?

**Comment by supermoos** [comment]
@tremby I think this issue may be related: https://github.com/gatsbyjs/gatsby/issues/7511#

**Comment by tremby** [comment]
@supermoos, are you asking for dimensions larger than the image actually is? If so, it probably is related.

**Comment by supermoos** [comment]
Nope, I'm not. The problem lies in the aspectRatio not being calculated on the actual image response from contentful, so it doesn't take into account what kind of crop mode is set.

**Comment by tremby** [comment]
Oh, in that case I'm not certain. But if the real dimensions are coming back from contentful (as it seems from Kyle's comments above that they should be) then it ought to be easy to fix. When I work on this one (hopefully soon), I'll look at your ticket too if someone else didn't already.

**Comment by supermoos** [comment]
Exactly, seems like simply a false assumption that it could be pre-calculated based on the request params was made. The dimensions should be easy to read off the response. Cool, that's amazing.

**Comment by tremby** [comment]
I just spent some time looking back at this issue.

The way I remembered it, the widths specified in `srcset` came out larger than the real widths of the images, which would definitely be a bug, but I'm not able to reproduce that. I think I may have misremembered, and it's simply the `img` element's `width` attribute which is coming out larger than the image's natural size.

With fresh eyes I'm actually not so sure any more that this is a bug. When the image is queried for with `fixed` (which used to be called `resolutions`), you specify the target width. The `srcset` coming back (if it does -- there isn't one if there's only one suitable image) is accurate. The width and height on the `img` element are what you asked for. (It occurs to me that maybe my change of mind has a lot to do with the renaming from `resolutions` to `fixed`!)

In the case this ticket is about, which is when the largest available image is *smaller* than the target dimensions, whether the behaviour above is what you want depends on the situation. You may want the image to max out at the natural size (as I did), or you may want those exact dimensions.

I'm not sure whether this should be an option, or it should be left alone in the knowledge that my workaround above (in the second comment here) works fine.

TLDR: @KyleAMathews, is it expected behaviour in the case where we query with `fixed` and ask for width 1000, and the image is only available up to 600 pixels wide, that the `img` tag has `width=1000`? If so, close. (The `srcset` attribute is fine.)

**Comment by supermoos** [comment]
@tremby The main problem I've seen is when you also define a crop mode, since some crop modes will return width/heights that cannot be precomputed without knowing the exact image dimensions of the source image.
